returnn/hdf.py

```diff
@@ -334,6 +343,13 @@ def run(self):
                         axis=0,
                     ).astype(self.output_dtype)
 
+                # compress if necessary
+                if self.compress_format is not None:
+                    out_sr = self.target_sampling_rate if self.target_sampling_rate else audio.samplerate
+                    buf = io.BytesIO()
+                    sf.write(buf, data, out_sr, format=self.compress_format)
+                    data = np.array([v for v in buf.getvalue()], dtype=np.uint8)
```

Comment:
is it required to use `uint8` here?
Would this ignore the `output_dtype` defined above, or would it rather be e.g. float data compressed and written in raw uint8 numbers?

Reply:
It is going to be the latter. After compression, the data is going to be bytes not matter what because it is stored in some sort of format (e.g. ogg). Making _that_ dtype configurable does not make much sense, I think. The output dtype still makes a difference because it limits the resolution of the audio samples before compression. After decompression you can again use any dtype you like, I don't think the compressed formats specify whether they store "floating point" or "int16" samples.

Does this answer your question?

Reply:
Yes, my question was answered, but then I went into a different direction researching what the correct dtype for byte arrays was (there is void, but I guess that is for byte-string-arrays?) and found `numpy.frombuffer` (would that be applicable and more efficient here?) and now I am wondering if this array of data is even meaningful in the context of the returnn hdf file format i.e. when we do 
```
                out_hdf.insert_batch(
                    inputs=data.reshape(1, -1, 1),
                    seq_len=[data.shape[0]],
                    seq_tag=[segment.fullname()],
                )
```
is the seq_len not completely wrong now? should we put the original seq_len?

Reply:
No, `seq_len` is correct. `seq_len` is if you insert a batch with shape [B,T,D], we want that `max(seq_len) = T`.

Reply:
> array of data is even meaningful in the context of the returnn hdf file format

To clarify: the data inside the HDF is mostly useless for training without a postprocessing step in between. This postprocessing step will go from "sequence of bytes that represents audio data in some arbitrary, possibly compressed container format" to "raw audio frames/PCM". As part of that, the seq length will go from "number of bytes that you need to decompress the audio in the container format" to the "correct" one for training, i.e. "number of raw PCM audio frames".

In the HDF the correct `seq_len` to use is "number of bytes that represent the audio in the given container format" because that number is needed to fetch the correct bytes from the HDF at runtime.

Reply:
ok, thanks for the detailed explanation. I think I now understand this implementation.

Now my only reservation is: "Isn't that a bit hacky and error prone? Should we not rather define a different dataset format that transparently handles this for the user?"

Imagine you have two data sources and only one of it is compressed... That will be a postprocessing function  / dataset  integration nightmare

Reply:
I think the required effort/overhead is the same if you have to use a different dataset type for one part of the data vs. having to wrap a part of the data in `PostprocessingDataset`. The only difference I could see is outfitting `HDFDataset` with transparent decompression functionality that it can apply entirely by itself by looking at e.g. an attribute in the HDF that says whether the data is compressed or raw audio.

Reply:
@albertz WDYT about a separate dataset implementation that uses the same disk format as the classic HDFDataset, but then transparently handles compressed audio by means of additional data keys that tell the dataset whether the data is audio and then whether it is compressed? I think importantly, this implementation would have/need to be backwards compatible (wrt. the disk format) so that existing HDF artefacts can be reused/used together without any special logic? Alternatively, the existing HDFDataset could be extended to support this use case. WDYT is the better option?

Reply:
I would not do any of that and just use it as you originally intended with `PostprocessingDataset`. I don't really see the big advantage. It sounds more like overengineering, and producing further problems (more code, more complexity in RETURNN).